File: Parallel-Csharp/src/Parallel/Linq101/MyLinq/TakeMethods.cs
using Parallel.Linq101.DataSources;

namespace Parallel.Linq101.MyLinq;

public class TakeMethods {
    public List<Product> GetProductList() => Products.ProductList;

    public List<Customer> GetCustomerList() => Customers.CustomerList;

    //original:4 target:45 form Partitions
    public List<int> IntTake() {
        int[] numbers = { 5, 4, 1, 3, 9, 8, 6, 7, 2, 0 };
        List<int> first3Numbers = numbers.Take(3).ToList();
        return first3Numbers;
    }

    public List<int> OddIntTake() {
        int[] numbers = { 5, 4, 1, 3, 9, 8, 6, 7, 2, 0 };
        List<int> first3Numbers = numbers.Where(i => i % 2 == 1).Take(3).ToList();
        return first3Numbers;
    }

    public List<int> IntSquareTake() {
        int[] numbers = { 5, 4, 1, 3, 9, 8, 6, 7, 2, 0 };
        List<int> first3Numbers = numbers.Select(i => i * i).Take(3).ToList();
        return first3Numbers;
    }

    public List<int> IntTakeCount(int n) {
        int[] numbers = { 5, 4, 1, 3, 9, 8, 6, 7, 2, 0 };
        List<int> first3Numbers = numbers.Take(n).ToList();
        return first3Numbers;
    }

    public List<int> IntTakeCount(int[] numbers, int n) {
        List<int> firstNNumbers = numbers.Take(n).ToList();
        return firstNNumbers;
    }

    public List<string> StringTake() {
        string[] words = { "cherry", "apple", "blueberry", "banana", "abacus" };
        List<string> firstNNumbers = words.Take(3).ToList();
        return firstNNumbers;
    }

    public List<string> LongStringTake() {
        string[] words = { "cherry", "apple", "blueberry", "banana", "abacus" };
        List<string> first3Words = words.Where(w => w.Length > 5).Take(3).ToList();
        return first3Words;
    }

    public List<int> StringLengthTake() {
        string[] words = { "cherry", "apple", "blueberry", "banana", "abacus" };
        List<int> first3Words = words.Select(w => w.Length).Take(3).ToList();
        return first3Words;
    }

    public List<string> StringTakeCount(int n) {
        string[] words = { "cherry", "apple", "blueberry", "banana", "abacus" };
        List<string> firstNWords = words.Take(n).ToList();
        return firstNWords;
    }

    public List<string> StringTakeCount(string[] strings, int n) {
        List<string> firstNWords = strings.Take(n).ToList();
        return firstNWords;
    }

    public List<Product> ProductTake() {
        List<Product> products = GetProductList();
        List<Product> first3Products = products.Take(3).ToList();
        return first3Products;
    }

    public List<Product> ExpensiveProductTake() {
        List<Product> products = GetProductList();
        List<Product> first3Products = products.Where(p => p.UnitPrice > 100).Take(3).ToList();
        return first3Products;
    }

    public List<Product> InStockProductTake() {
        List<Product> products = GetProductList();
        List<Product> first3Products = products.Where(p => p.UnitsInStock > 0).Take(3).ToList();
        return first3Products;
    }

    public Dictionary<string, List<Product>> ProductGroupTake() {
        List<Product> products = GetProductList();
        Dictionary<string, List<Product>> first3Groups = products.GroupBy(p => p.Category)
            .Take(3).ToDictionary(g => g.Key, g => g.ToList());
        return first3Groups;
    }

    public Dictionary<string, List<Product>> ProductEachGroupTake() {
        List<Product> products = GetProductList();
        Dictionary<string, List<Product>> first3EachGroups = products.GroupBy(p => p.Category)
            .ToDictionary(g => g.Key, g => g.Take(3).ToList());
        return first3EachGroups;
    }

    public List<Order> CustomerOrdersTake() {
        List<Customer> customers = GetCustomerList();
        List<Order> first3Orders = customers.SelectMany(c => c.Orders)
            .Take(3).ToList();
        return first3Orders;
    }

    public List<Tuple<string, DateTime>> CustomerOrdersNestedTake() {
        List<Customer> customers = GetCustomerList();
        List<Tuple<string, DateTime>> first3Orders = customers.SelectMany(c => c.Orders, (c, o) => Tuple.Create(
            c.CustomerId, o.OrderDate)).Take(3).ToList();
        return first3Orders;
    }

    public List<Tuple<string, DateTime>> CustomerOrdersEachNestedTake() {
        List<Customer> customers = GetCustomerList();
        List<Tuple<string, DateTime>> first3EachOrders = customers
            .SelectMany(c => c.Orders.Take(3), (c, o) => Tuple.Create(
                c.CustomerId, o.OrderDate)).ToList();
        return first3EachOrders;
    }

    public List<Tuple<string, DateTime>> WaCustomerOrdersNestedTake() {
        List<Customer> customers = GetCustomerList();
        List<Tuple<string, DateTime>> first3WAOrders = customers.Where(c => c.Region == "WA")
            .SelectMany(c => c.Orders, (c, o) => Tuple.Create(
                c.CustomerId, o.OrderDate)).Take(3).ToList();
        return first3WAOrders;
    }

    public List<Tuple<string, DateTime>> WaCustomerOrdersEachNestedTake() {
        List<Customer> customers = GetCustomerList();
        List<Tuple<string, DateTime>> first3EachWAOrders = customers.Where(c => c.Region == "WA")
            .SelectMany(c => c.Orders.Take(3), (c, o) => Tuple.Create(
                c.CustomerId, o.OrderDate)).ToList();
        return first3EachWAOrders;
    }

    public List<int> IntTakeWhile() {
        int[] numbers = { 5, 4, 1, 3, 9, 8, 6, 7, 2, 0 };
        List<int> firstNumbersWhile = numbers.TakeWhile(i => i > 3).ToList();
        return firstNumbersWhile;
    }

    public List<int> OddIntTakeWhile() {
        int[] numbers = { 5, 4, 1, 3, 9, 8, 6, 7, 2, 0 };
        List<int> firstNumbersWhile = numbers.Where(i => i % 2 == 1).TakeWhile(i => i > 3).ToList();
        return firstNumbersWhile;
    }

    public List<int> IntSquareTakeWhile() {
        int[] numbers = { 5, 4, 1, 3, 9, 8, 6, 7, 2, 0 };
        List<int> firstNumbersWhile = numbers.Select(i => i * i).TakeWhile(i => i > 3).ToList();
        return firstNumbersWhile;
    }

    public List<int> IntTakeWhileCount(int n) {
        int[] numbers = { 5, 4, 1, 3, 9, 8, 6, 7, 2, 0 };
        List<int> firstNumbersWhile = numbers.TakeWhile(i => i > n).ToList();
        return firstNumbersWhile;
    }

    public List<int> IntTakeWhileCount(int[] numbers, int n) {
        List<int> firstNumbersWhile = numbers.TakeWhile(i => i > n).ToList();
        return firstNumbersWhile;
    }

    public List<string> StringTakeWhile() {
        string[] words = { "cherry", "apple", "blueberry", "banana", "abacus" };
        List<string> firstWordsWhile = words.TakeWhile(w => w[0] > 'b').ToList();
        return firstWordsWhile;
    }

    public List<string> LongStringTakeWhile() {
        string[] words = { "cherry", "apple", "blueberry", "banana", "abacus" };
        List<string> firstWordsWhile = words.Where(w => w.Length > 5).TakeWhile(w => w[0] > 'b').ToList();
        return firstWordsWhile;
    }

    public List<int> StringLengthTakeWhile() {
        string[] words = { "cherry", "apple", "blueberry", "banana", "abacus" };
        List<int> firstWordsWhile = words.Select(w => w.Length).TakeWhile(n => n > 5).ToList();
        return firstWordsWhile;
    }

    public List<string> StringTakeWhileCount(char c) {
        string[] words = { "cherry", "apple", "blueberry", "banana", "abacus" };
        List<string> firstWordsWhile = words.TakeWhile(w => w[0] > c).ToList();
        return firstWordsWhile;
    }

    public List<string> StringTakeWhileCount(string[] strings, char c) {
        List<string> firstWordsWhile = strings.TakeWhile(w => w[0] > c).ToList();
        return firstWordsWhile;
    }

    public List<Product> ProductTakeWhile() {
        List<Product> products = GetProductList();
        List<Product> firstProductsWhile = products.TakeWhile(p => p.ProductId > 100).ToList();
        return firstProductsWhile;
    }

    public List<Product> ExpensiveProductTakeWhile() {
        List<Product> products = GetProductList();
        List<Product> firstProductsWhile =
            products.Where(p => p.UnitPrice > 100).TakeWhile(p => p.ProductId > 100).ToList();
        return firstProductsWhile;
    }

    public List<Product> InStockProductTakeWhile() {
        List<Product> products = GetProductList();
        List<Product> firstProductsWhile =
            products.Where(p => p.UnitsInStock > 0).TakeWhile(p => p.ProductId > 100).ToList();
        return firstProductsWhile;
    }

    public Dictionary<string, List<Product>> ProductGroupTakeWhile() {
        List<Product> products = GetProductList();
        Dictionary<string, List<Product>> firstGroupsWhile = products.GroupBy(p => p.Category)
            .TakeWhile(g => g.Key.Length > 6).ToDictionary(g => g.Key, g => g.ToList());
        return firstGroupsWhile;
    }

    public Dictionary<string, List<Product>> ProductEachGroupTakeWhile() {
        List<Product> products = GetProductList();
        Dictionary<string, List<Product>> firstEachGroupsWhile = products.GroupBy(p => p.Category)
            .ToDictionary(g => g.Key, g => g.TakeWhile(p => p.ProductId > 100).ToList());
        return firstEachGroupsWhile;
    }

    public List<Order> CustomerOrdersTakeWhile() {
        List<Customer> customers = GetCustomerList();
        List<Order> firstOrdersWhile = customers.SelectMany(c => c.Orders)
            .TakeWhile(o => o.OrderDate > new DateTime(2000, 1, 1)).ToList();
        return firstOrdersWhile;
    }

    public List<Tuple<string, DateTime>> CustomerOrdersNestedTakeWhile() {
        List<Customer> customers = GetCustomerList();
        List<Tuple<string, DateTime>> firstOrdersWhile = customers.SelectMany(c => c.Orders, (c, o) => Tuple.Create(
            c.CustomerId, o.OrderDate)).TakeWhile(t => t.Item2 > new DateTime(
            2000, 1, 1)).ToList();
        return firstOrdersWhile;
    }

    public List<Tuple<string, DateTime>> CustomerOrdersEachNestedTakeWhile() {
        List<Customer> customers = GetCustomerList();
        List<Tuple<string, DateTime>> firstOrdersEachWhile = customers
            .SelectMany(c => c.Orders.TakeWhile(o => o.OrderDate > new DateTime(
                2000, 1, 1)), (c, o) => Tuple.Create(
                c.CustomerId, o.OrderDate)).ToList();
        return firstOrdersEachWhile;
    }

    public List<Tuple<string, DateTime>> WaCustomerOrdersNestedTakeWhile() {
        List<Customer> customers = GetCustomerList();
        List<Tuple<string, DateTime>> firstOrdersWhile = customers.Where(c => c.Region == "WA")
            .SelectMany(c => c.Orders, (c, o) => Tuple.Create(
                c.CustomerId, o.OrderDate)).TakeWhile(t => t.Item2 > new DateTime(
                2000, 1, 1)).ToList();
        return firstOrdersWhile;
    }

    public List<Tuple<string, DateTime>> WaCustomerOrdersEachNestedTakeWhile() {
        List<Customer> customers = GetCustomerList();
        List<Tuple<string, DateTime>> firstOrdersEachWhile = customers.Where(c => c.Region == "WA")
            .SelectMany(c => c.Orders.TakeWhile(o => o.OrderDate > new DateTime(
                2000, 1, 1)), (c, o) => Tuple.Create(
                c.CustomerId, o.OrderDate)).ToList();
        return firstOrdersEachWhile;
    }

    public List<int> IntTakeWhileIndexed() {
        int[] numbers = { 5, 4, 1, 3, 9, 8, 6, 7, 2, 0 };
        List<int> firstNumbers = numbers.TakeWhile((n, index) => n > index).ToList();
        return firstNumbers;
    }
    public List<int> IntSquareTakeWhileIndexed() {
        int[] numbers = { 5, 4, 1, 3, 9, 8, 6, 7, 2, 0 };
        List<int> firstNumbers = numbers.Select(i => i * i).TakeWhile((n, index) => n > index).ToList();
        return firstNumbers;
    }
    
    public List<double> IntSqrtSkipWhileIndexed() {
        int[] numbers = { 5, 4, 1, 3, 9, 8, 6, 7, 2, 0 };
        List<double> afterNumbers = numbers.Select(i => Math.Sqrt(i)).TakeWhile((n, index) => n > index).ToList();
        return afterNumbers;
    }

    public List<int> IntTakeWhileIndexedCount(int n) {
        int[] numbers = { 5, 4, 1, 3, 9, 8, 6, 7, 2, 0 };
        List<int> firstNumbers = numbers.TakeWhile((n, index) => n > index).ToList();
        return firstNumbers;
    }

    public List<int> IntTakeWhileIndexedCount(int[] numbers, int n) {
        List<int> firstNumbers = numbers.TakeWhile((n, index) => n > index).ToList();
        return firstNumbers;
    }
}

File: Parallel-Java/src/parallel/linq101/mylinq/TakeMethods.java
package parallel.linq101.mylinq;

import org.javatuples.Pair;
import parallel.linq101.datasource.Customer;
import parallel.linq101.datasource.Data;
import parallel.linq101.datasource.Order;
import parallel.linq101.datasource.Product;

import java.time.LocalDate;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.stream.IntStream;

import static java.util.stream.Collectors.groupingBy;
import static java.util.stream.Collectors.toMap;

public class TakeMethods {
    public List<Product> getProductList() {
        return Data.getProductList();
    }

    public List<Customer> getCustomerList() {
        return Data.getCustomerList();
    }

    public List<Integer> intTake() {
        int[] numbers = {5, 4, 1, 3, 9, 8, 6, 7, 2, 0};
        List<Integer> after3Numbers = Arrays.stream(numbers).limit(3).boxed().toList();
        return after3Numbers;
    }

    public List<Integer> oddIntTake() {
        int[] numbers = {5, 4, 1, 3, 9, 8, 6, 7, 2, 0};
        List<Integer> after3Numbers = Arrays.stream(numbers).filter(i -> i % 2 == 1).limit(3).boxed().toList();
        return after3Numbers;
    }

    public List<Integer> intSquareTake() {
        int[] numbers = {5, 4, 1, 3, 9, 8, 6, 7, 2, 0};
        List<Integer> after3Numbers = Arrays.stream(numbers).map(i -> i * i).limit(3).boxed().toList();
        return after3Numbers;
    }

    public List<Integer> intTakeCount(int n) {
        int[] numbers = {5, 4, 1, 3, 9, 8, 6, 7, 2, 0};
        List<Integer> after3Numbers = Arrays.stream(numbers).limit(n).boxed().toList();
        return after3Numbers;
    }

    public List<Integer> intTakeCount(int[] numbers, int n) {
        List<Integer> afterNNumbers = Arrays.stream(numbers).limit(n).boxed().toList();
        return afterNNumbers;
    }

    public List<String> stringTake() {
        String[] words = {"cherry", "apple", "blueberry", "banana", "abacus"};
        List<String> afterNNumbers = Arrays.stream(words).limit(3).toList();
        return afterNNumbers;
    }

    public List<String> longStringTake() {
        String[] words = {"cherry", "apple", "blueberry", "banana", "abacus"};
        List<String> after3Words = Arrays.stream(words).filter(w -> w.length() > 5).limit(3).toList();
        return after3Words;
    }

    public List<Integer> stringLengthTake() {
        String[] words = {"cherry", "apple", "blueberry", "banana", "abacus"};
        List<Integer> after3Words = Arrays.stream(words).map(w -> w.length()).limit(3).toList();
        return after3Words;
    }

    public List<String> stringTakeCount(int n) {
        String[] words = {"cherry", "apple", "blueberry", "banana", "abacus"};
        List<String> afterNWords = Arrays.stream(words).limit(3).toList();
        return afterNWords;
    }

    public List<String> stringTakeCount(String[] strings, int n) {
        List<String> afterNWords = Arrays.stream(strings).limit(3).toList();
        return afterNWords;
    }

    public List<Product> productTake() {
        List<Product> products = getProductList();
        List<Product> after3Products = products.stream().limit(3).toList();
        return after3Products;
    }

    public List<Product> expensiveProductTake() {
        List<Product> products = getProductList();
        List<Product> after3Products = products.stream().filter(p -> p.getUnitPrice() > 100).limit(3).toList();
        return after3Products;
    }

    public List<Product> inStockProductTake() {
        List<Product> products = getProductList();
        List<Product> after3Products = products.stream().filter(p -> p.getUnitsInStock() > 100).limit(3).toList();
        return after3Products;
    }

    public Map<String, List<Product>> productGroupTake() {
        List<Product> products = getProductList();
        Map<String, List<Product>> after3Groups = products.stream().collect(groupingBy(p -> p.getCategory()))
                .entrySet().stream().limit(3).collect(toMap(g -> g.getKey(), g -> g.getValue()));
        return after3Groups;
    }

    public Map<String, List<Product>> productEachGroupTake() {
        List<Product> products = getProductList();
        Map<String, List<Product>> after3EachGroups = products.stream().collect(groupingBy(p -> p.getCategory()))
                .entrySet().stream().collect(toMap(g -> g.getKey(), g -> g.getValue().stream().limit(3).toList()));
        return after3EachGroups;
    }

    public List<Order> customerOrdersTake() {
        List<Customer> customers = getCustomerList();
        List<Order> after3Orders = customers.stream().flatMap(c -> c.getOrders().stream()).limit(3).toList();
        return after3Orders;
    }

    public List<Pair<String, LocalDate>> customerOrdersNestedTake() {
        List<Customer> customers = getCustomerList();
        List<Pair<String, LocalDate>> after3Orders = customers.stream().flatMap(c -> c.getOrders().stream()
                .map(o -> Pair.with(c.getCustomerId(), o.getOrderDate()))).limit(3).toList();
        return after3Orders;
    }

    public List<Pair<String, LocalDate>> customerOrdersEachNestedTake() {
        List<Customer> customers = getCustomerList();
        List<Pair<String, LocalDate>> after3EachOrders = customers.stream().flatMap(c -> c.getOrders().stream().limit(3)
                .map(o -> Pair.with(c.getCustomerId(), o.getOrderDate()))).toList();
        return after3EachOrders;
    }

    public List<Pair<String, LocalDate>> waCustomerOrdersNestedTake() {
        List<Customer> customers = getCustomerList();
        List<Pair<String, LocalDate>> after3WAOrders = customers.stream().filter(c -> c.getRegion().equals("WA"))
                .flatMap(c -> c.getOrders().stream().map(o ->
                        Pair.with(c.getCustomerId(), o.getOrderDate()))).limit(3).toList();
        return after3WAOrders;
    }

    public List<Pair<String, LocalDate>> waCustomerOrdersEachNestedTake() {
        List<Customer> customers = getCustomerList();
        List<Pair<String, LocalDate>> after3EachWAOrders = customers.stream().filter(c -> c.getRegion().equals("WA"))
                .flatMap(c -> c.getOrders().stream().limit(3).map(o ->
                        Pair.with(c.getCustomerId(), o.getOrderDate()))).toList();
        return after3EachWAOrders;
    }

    public List<Integer> intTakeWhile() {
        int[] numbers = {5, 4, 1, 3, 9, 8, 6, 7, 2, 0};
        List<Integer> after3Numbers = Arrays.stream(numbers).takeWhile(i -> i > 3).boxed().toList();
        return after3Numbers;
    }

    public List<Integer> oddIntTakeWhile() {
        int[] numbers = {5, 4, 1, 3, 9, 8, 6, 7, 2, 0};
        List<Integer> after3Numbers = Arrays.stream(numbers).filter(i -> i % 2 == 1).takeWhile(i -> i > 3).boxed().toList();
        return after3Numbers;
    }

    public List<Integer> intSquareTakeWhile() {
        int[] numbers = {5, 4, 1, 3, 9, 8, 6, 7, 2, 0};
        List<Integer> after3Numbers = Arrays.stream(numbers).map(i -> i * i).takeWhile(i -> i > 3).boxed().toList();
        return after3Numbers;
    }

    public List<Integer> intTakeWhileCount(int n) {
        int[] numbers = {5, 4, 1, 3, 9, 8, 6, 7, 2, 0};
        List<Integer> after3Numbers = Arrays.stream(numbers).takeWhile(i -> i > n).boxed().toList();
        return after3Numbers;
    }

    public List<Integer> intTakeWhileCount(int[] numbers, int n) {
        List<Integer> afterNNumbers = Arrays.stream(numbers).takeWhile(i -> i > n).boxed().toList();
        return afterNNumbers;
    }

    public List<String> stringTakeWhile() {
        String[] words = {"cherry", "apple", "blueberry", "banana", "abacus"};
        List<String> afterNNumbers = Arrays.stream(words).takeWhile(w -> w.charAt(0) > 'b').toList();
        return afterNNumbers;
    }

    public List<String> longStringTakeWhile() {
        String[] words = {"cherry", "apple", "blueberry", "banana", "abacus"};
        List<String> after3Words = Arrays.stream(words).filter(w -> w.length() > 5).takeWhile(w -> w.charAt(0) > 'b').toList();
        return after3Words;
    }

    public List<Integer> stringLengthTakeWhile() {
        String[] words = {"cherry", "apple", "blueberry", "banana", "abacus"};
        List<Integer> after3Words = Arrays.stream(words).map(w -> w.length()).takeWhile(n -> n > 5).toList();
        return after3Words;
    }

    public List<String> stringTakeWhileCount(char c) {
        String[] words = {"cherry", "apple", "blueberry", "banana", "abacus"};
        List<String> afterNWords = Arrays.stream(words).takeWhile(w -> w.charAt(0) > c).toList();
        return afterNWords;
    }

    public List<String> stringTakeWhileCount(String[] strings, char c) {
        List<String> afterNWords = Arrays.stream(strings).takeWhile(w -> w.charAt(0) > c).toList();
        return afterNWords;
    }

    public List<Product> productTakeWhile() {
        List<Product> products = getProductList();
        List<Product> after3Products = products.stream().takeWhile(p -> p.getProductId() > 100).toList();
        return after3Products;
    }

    public List<Product> expensiveProductTakeWhile() {
        List<Product> products = getProductList();
        List<Product> after3Products = products.stream().filter(p -> p.getUnitPrice() > 100).takeWhile(p -> p.getProductId() > 100).toList();
        return after3Products;
    }

    public List<Product> inStockProductTakeWhile() {
        List<Product> products = getProductList();
        List<Product> after3Products = products.stream().filter(p -> p.getUnitsInStock() > 100).takeWhile(p -> p.getProductId() > 100).toList();
        return after3Products;
    }

    public Map<String, List<Product>> productGroupTakeWhile() {
        List<Product> products = getProductList();
        Map<String, List<Product>> after3Groups = products.stream().collect(groupingBy(p -> p.getCategory()))
                .entrySet().stream().takeWhile(g -> g.getKey().length() > 6).collect(toMap(g -> g.getKey(), g -> g.getValue()));
        return after3Groups;
    }

    public Map<String, List<Product>> productEachGroupTakeWhile() {
        List<Product> products = getProductList();
        Map<String, List<Product>> after3EachGroups = products.stream().collect(groupingBy(p -> p.getCategory()))
                .entrySet().stream().collect(toMap(g -> g.getKey(), g -> g.getValue().stream().takeWhile(p -> p.getProductId() > 100).toList()));
        return after3EachGroups;
    }

    public List<Order> customerOrdersTakeWhile() {
        List<Customer> customers = getCustomerList();
        List<Order> after3Orders = customers.stream().flatMap(c -> c.getOrders().stream())
                .takeWhile(o -> o.getOrderDate().isAfter(LocalDate.of(2000, 1, 1))).toList();
        return after3Orders;
    }

    public List<Pair<String, LocalDate>> customerOrdersNestedTakeWhile() {
        List<Customer> customers = getCustomerList();
        List<Pair<String, LocalDate>> after3Orders = customers.stream().flatMap(c -> c.getOrders().stream()
                        .map(o -> Pair.with(c.getCustomerId(), o.getOrderDate())))
                .takeWhile(t -> t.getValue1().isAfter(LocalDate.of(2000, 1, 1))).toList();
        return after3Orders;
    }

    public List<Pair<String, LocalDate>> customerOrdersEachNestedTakeWhile() {
        List<Customer> customers = getCustomerList();
        List<Pair<String, LocalDate>> after3EachOrders = customers.stream().flatMap(c -> c.getOrders().stream()
                .takeWhile(o -> o.getOrderDate().isAfter(LocalDate.of(2000, 1, 1)))
                .map(o -> Pair.with(c.getCustomerId(), o.getOrderDate()))).toList();
        return after3EachOrders;
    }

    public List<Pair<String, LocalDate>> waCustomerOrdersNestedTakeWhile() {
        List<Customer> customers = getCustomerList();
        List<Pair<String, LocalDate>> after3WAOrders = customers.stream().filter(c -> c.getRegion().equals("WA"))
                .flatMap(c -> c.getOrders().stream().map(o -> Pair.with(c.getCustomerId(), o.getOrderDate())))
                .takeWhile(t -> t.getValue1().isAfter(LocalDate.of(2000, 1, 1))).toList();
        return after3WAOrders;
    }

    public List<Pair<String, LocalDate>> waCustomerOrdersEachNestedTakeWhile() {
        List<Customer> customers = getCustomerList();
        List<Pair<String, LocalDate>> after3EachWAOrders = customers.stream().filter(c -> c.getRegion().equals("WA"))
                .flatMap(c -> c.getOrders().stream().takeWhile(o -> o.getOrderDate()
                                .isAfter(LocalDate.of(2000, 1, 1)))
                        .map(o -> Pair.with(c.getCustomerId(), o.getOrderDate()))).toList();
        return after3EachWAOrders;
    }

    public List<Integer> intTakeWhileIndexed() {
        int[] numbers = {5, 4, 1, 3, 9, 8, 6, 7, 2, 0};
        List<Integer> afterNumbers = IntStream.range(0, numbers.length)
                .takeWhile(i -> numbers[i] > i).map(i -> numbers[i]).boxed().toList();
        return afterNumbers;
    }

    public List<Integer> intSquareTakeWhileIndexed() {
        int[] numbers = {5, 4, 1, 3, 9, 8, 6, 7, 2, 0};
        List<Integer> afterNumbers =  IntStream.range(0, numbers.length)
                .takeWhile(i -> numbers[i]*numbers[i] > i).map(i -> numbers[i]*numbers[i]).boxed().toList();
        return afterNumbers;
    }

    public List<Double> intSqrtTakeWhileIndexed() {
        int[] numbers = {5, 4, 1, 3, 9, 8, 6, 7, 2, 0};
        List<Double> afterNumbers =  IntStream.range(0, numbers.length)
                .takeWhile(i -> Math.sqrt(numbers[i]) > i).mapToDouble(i -> Math.sqrt(numbers[i]) ).boxed().toList();
        return afterNumbers;
    }

    public List<Integer> intTakeWhileIndexedCount(int n) {
        int[] numbers = {5, 4, 1, 3, 9, 8, 6, 7, 2, 0};
        List<Integer> afterNumbers = IntStream.range(0, numbers.length)
                .takeWhile(i -> n > i).map(i -> numbers[i]).boxed().toList();
        return afterNumbers;
    }

    public List<Integer> intTakeWhileIndexedCount(int[] numbers, int n) {
        List<Integer> afterNumbers = IntStream.range(0, numbers.length)
                .takeWhile(i -> n > i).map(i -> numbers[i]).boxed().toList();
        return afterNumbers;
    }

}
